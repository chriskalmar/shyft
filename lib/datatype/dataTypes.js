
import DataType, {
  DataTypeUser,
} from './DataType';



export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user'
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs'
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values'
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values'
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values'
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values'
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values'
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects'
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp'
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information'
})
