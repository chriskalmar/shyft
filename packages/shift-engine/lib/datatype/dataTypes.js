
import DataType from './DataType';
import DataTypeUser from './DataTypeUser';

import casual from 'casual';
import _ from 'lodash';


export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})



export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(-2^10, 2^10),
})


export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  mock: /* istanbul ignore next */
    () => casual.integer(2^20, 2^51).toString(),
})


export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  mock: /* istanbul ignore next */
    () => casual.double(-2^10, 2^10),
})


export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  mock: /* istanbul ignore next */
    () => casual.boolean,
})


export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  mock: /* istanbul ignore next */
    () => casual.title,
})


export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  mock: /* istanbul ignore next */
    randomJson,
})


export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  mock: /* istanbul ignore next */
    () => new Date(casual.unix_time * 1000),
})


/* istanbul ignore next */
function randomJson() {
  const ret = {}
  const count = casual.integer(1, 5)

  _.times(count, () => {
    const key = _.camelCase( casual.words(2) )
    const value = Math.random() > 0.5
      ? casual.title
      : casual.integer

    ret[ key ] = value
  })

  return ret
}
