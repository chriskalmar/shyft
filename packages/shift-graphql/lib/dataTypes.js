
import GraphQLJSON from 'graphql-type-json';

import {
  GraphQLScalarType,
  Kind,
  GraphQLError,
} from 'graphql';


export {
  GraphQLJSON,
}


export const GraphQLBigInt = new GraphQLScalarType({
  name: 'BigInt',
  description:
    'The `BigInt` scalar type represents a 64-bit integer. As JavaScript ' +
    'is limited to a precision of 53 bits on integer values, all `BigInt` ' +
    'values will be output as strings.',
  serialize: String,
  parseValue: String,
  parseLiteral(ast) {
    return ast.kind === Kind.STRING ? ast.value : null;
  }
});


export const GraphQLDateTime = new GraphQLScalarType({
  name: 'DateTime',
  description:
    'The `DateTime` scalar type represents a date and time string.',
  serialize: (value) => {
    return value
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError('Query error: Can only parse strings but got a: ' + ast.kind, [ ast ]);
    }

    if (isNaN(Date.parse(ast.value))) {
      throw new GraphQLError('Query error: String is not a valid date time string', [ ast ]);
    }

    return ast.value
  }
});

