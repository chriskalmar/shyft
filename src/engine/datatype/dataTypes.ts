import { DataType } from './DataType';
import { DataTypeUser } from './DataTypeUser';
import { randomJson } from '../util';
import * as casual from 'casual';

export const DataTypeUserID = new DataTypeUser({
  name: 'DataTypeUserID',
  description: 'Data type representing a reference to a user',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeID = new DataType({
  name: 'DataTypeID',
  description: 'Data type representing unique IDs',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeInteger = new DataType({
  name: 'DataTypeInteger',
  description: 'Data type representing integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBigInt = new DataType({
  name: 'DataTypeBigInt',
  description: 'Data type representing big integer values',
  /* istanbul ignore next */
  mock: () => casual.integer(2 ^ 20, 2 ^ 51).toString(),
});

export const DataTypeFloat = new DataType({
  name: 'DataTypeFloat',
  description: 'Data type representing float values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeDouble = new DataType({
  name: 'DataTypeDouble',
  description: 'Data type representing double precision values',
  /* istanbul ignore next */
  mock: () => casual.double(-2 ^ 10, 2 ^ 10),
});

export const DataTypeBoolean = new DataType({
  name: 'DataTypeBoolean',
  description: 'Data type representing boolean values',
  /* istanbul ignore next */
  mock: () => casual.boolean,
  enforceRequired: true,
  defaultValue() {
    return false;
  },
});

export const DataTypeString = new DataType({
  name: 'DataTypeString',
  description: 'Data type representing text values',
  /* istanbul ignore next */
  mock: () => casual.title,
});

export const DataTypeJson = new DataType({
  name: 'DataTypeJson',
  description: 'Data type representing json objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeTimestamp = new DataType({
  name: 'DataTypeTimestamp',
  description: 'Data type representing a timestamp',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimestampTz = new DataType({
  name: 'DataTypeTimestampTz',
  description: 'Data type representing a timestamp with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeDate = new DataType({
  name: 'DataTypeDate',
  description: 'Data type representing a date',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTime = new DataType({
  name: 'DataTypeTime',
  description: 'Data type representing time',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeTimeTz = new DataType({
  name: 'DataTypeTimeTz',
  description: 'Data type representing time with time zone information',
  /* istanbul ignore next */
  mock: () => new Date(casual.unix_time * 1000),
});

export const DataTypeUUID = new DataType({
  name: 'DataTypeUUID',
  description: 'Data type representing a UUID',
  /* istanbul ignore next */
  mock: () => casual.uuid,
});

export const DataTypeI18n = new DataType({
  name: 'DataTypeI18n',
  description: 'Data type representing translation objects',
  /* istanbul ignore next */
  mock: randomJson,
});

export const DataTypeUpload = new DataType({
  name: 'DataTypeUpload',
  description: 'Data type representing a file upload',
  mock: () => null,
});
