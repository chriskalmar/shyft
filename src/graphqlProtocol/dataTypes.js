import { GraphQLScalarType, Kind, GraphQLError } from 'graphql';
import { serializeCursor, deserializeCursor } from './util';
import * as dateFns from 'date-fns';

const dateTimeRegex = /^\d+-[0-9]{2}-[0-9]{2}[\sT]([0-1][0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\s?[+-]([0-1][0-9]|2[0-3])(:[0-5][0-9])?)?$/;
const timeRegex = /^([0-1][0-9]|2[0-3]):[0-5][0-9](:[0-5][0-9])?(\s?[+-]([0-1][0-9]|2[0-3])(:[0-5][0-9])?)?$/;

export { GraphQLJSON } from 'graphql-type-json';

export const GraphQLCursor = new GraphQLScalarType({
  name: 'Cursor',
  description:
    'A cursor to define a location in a data set used for pagination.',
  serialize: serializeCursor,
  parseValue: deserializeCursor,
  parseLiteral(ast) {
    return ast.kind === Kind.STRING ? deserializeCursor(ast.value) : null;
  },
});

export const GraphQLBigInt = new GraphQLScalarType({
  name: 'BigInt',
  description:
    'The `BigInt` scalar type represents a 64-bit integer. As JavaScript ' +
    'is limited to a precision of 53 bits on integer values, all `BigInt` ' +
    'values will be output as strings.',
  serialize: String,
  parseValue: String,
  parseLiteral(ast) {
    return ast.kind === Kind.STRING ? ast.value : null;
  },
});

export const GraphQLDateTime = new GraphQLScalarType({
  name: 'DateTime',
  description: 'The `DateTime` scalar type represents a date and time string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    if (!dateTimeRegex.test(ast.value)) {
      throw new GraphQLError(
        'Query error: String is not a valid date time string',
        [ ast ],
      );
    }

    const dateString = ast.value.substring(0, 10);
    const asDate = dateFns.parse(dateString);

    if (
      !dateFns.isValid(asDate) ||
      dateString !== dateFns.format(asDate, 'YYYY-MM-DD')
    ) {
      throw new GraphQLError(
        'Query error: String is not a valid date time string',
        [ ast ],
      );
    }

    return ast.value;
  },
});

export const GraphQLDate = new GraphQLScalarType({
  name: 'Date',
  description: 'The `Date` scalar type represents a date string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    const asDate = dateFns.parse(ast.value);

    if (
      !dateFns.isValid(asDate) ||
      ast.value !== dateFns.format(asDate, 'YYYY-MM-DD')
    ) {
      throw new GraphQLError('Query error: String is not a valid date string', [
        ast,
      ]);
    }

    return ast.value;
  },
});

export const GraphQLTime = new GraphQLScalarType({
  name: 'Time',
  description: 'The `Time` scalar type represents a time string.',
  serialize: value => {
    return value;
  },
  parseValue: String,
  parseLiteral(ast) {
    if (ast.kind !== Kind.STRING) {
      throw new GraphQLError(
        `Query error: Can only parse strings but got a: ${ast.kind}`,
        [ ast ],
      );
    }

    if (!timeRegex.test(ast.value)) {
      throw new GraphQLError('Query error: String is not a valid time string', [
        ast,
      ]);
    }

    return ast.value;
  },
});
