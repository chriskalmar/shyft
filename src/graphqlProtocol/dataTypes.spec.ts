/* eslint-disable @typescript-eslint/explicit-function-return-type */

import { GraphQLDateTime, GraphQLDate, GraphQLTime } from './dataTypes';
import { parseValue } from 'graphql';

describe('dataTypes', () => {
  it('GraphQLDateTime', () => {
    const bad1 = () => GraphQLDateTime.parseLiteral(parseValue('"foo"'), null);
    expect(bad1).toThrowErrorMatchingSnapshot();

    const bad2 = () =>
      GraphQLDateTime.parseLiteral(
        parseValue('"2000-01-02 111:12:13+14:15"'),
        null,
      );
    expect(bad2).toThrowErrorMatchingSnapshot('bad hour');

    const bad3 = () =>
      GraphQLDateTime.parseLiteral(
        parseValue('"2000-31-02 11:12:13+14:15"'),
        null,
      );
    expect(bad3).toThrowErrorMatchingSnapshot('bad month');

    const bad4 = () =>
      GraphQLDateTime.parseLiteral(
        parseValue('"2000-01-02 11:62:13+14:15"'),
        null,
      );
    expect(bad4).toThrowErrorMatchingSnapshot('bad minute');

    const good1 = () =>
      GraphQLDateTime.parseLiteral(
        parseValue('"2000-01-02 11:12:13+14:15"'),
        null,
      );
    expect(good1).not.toThrow();

    const good2 = () =>
      GraphQLDateTime.parseLiteral(
        parseValue('"0003-01-02 21:12:13+14:15"'),
        null,
      );
    expect(good2).not.toThrow();
  });

  it('GraphQLDate', () => {
    const bad1 = () => GraphQLDate.parseLiteral(parseValue('"foo"'), null);
    expect(bad1).toThrowErrorMatchingSnapshot();

    const bad2 = () =>
      GraphQLDate.parseLiteral(parseValue('"2000-01-32"'), null);
    expect(bad2).toThrowErrorMatchingSnapshot('bad day');

    const bad3 = () =>
      GraphQLDate.parseLiteral(parseValue('"2000-31-02"'), null);
    expect(bad3).toThrowErrorMatchingSnapshot('bad month');

    const bad4 = () => GraphQLDate.parseLiteral(parseValue('"01-02"'), null);
    expect(bad4).toThrowErrorMatchingSnapshot('bad date');

    const good1 = () =>
      GraphQLDate.parseLiteral(parseValue('"2000-01-02"'), null);
    expect(good1).not.toThrow();

    const good2 = () =>
      GraphQLDate.parseLiteral(parseValue('"0003-01-02"'), null);
    expect(good2).not.toThrow();
  });

  it('GraphQLTime', () => {
    const bad1 = () => GraphQLTime.parseLiteral(parseValue('"foo"'), null);
    expect(bad1).toThrowErrorMatchingSnapshot();

    const bad2 = () =>
      GraphQLTime.parseLiteral(parseValue('"111:12:13+14:15"'), null);
    expect(bad2).toThrowErrorMatchingSnapshot('bad hour');

    const bad3 = () =>
      GraphQLTime.parseLiteral(parseValue('"11:12:13+24:15"'), null);
    expect(bad3).toThrowErrorMatchingSnapshot('bad tz hour');

    const bad4 = () =>
      GraphQLTime.parseLiteral(parseValue('"11:62:13+14:15"'), null);
    expect(bad4).toThrowErrorMatchingSnapshot('bad minute');

    const good1 = () =>
      GraphQLTime.parseLiteral(parseValue('"11:12:13+14:15"'), null);
    expect(good1).not.toThrow();

    const good2 = () =>
      GraphQLTime.parseLiteral(parseValue('"21:12:13"'), null);
    expect(good2).not.toThrow();

    const good3 = () => GraphQLTime.parseLiteral(parseValue('"21:12"'), null);
    expect(good3).not.toThrow();
  });
});
