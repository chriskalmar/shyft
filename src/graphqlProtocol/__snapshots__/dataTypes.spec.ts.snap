// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataTypes GraphQLDate 1`] = `"Query error: String is not a valid date string"`;

exports[`dataTypes GraphQLDate: bad date 1`] = `"Query error: String is not a valid date string"`;

exports[`dataTypes GraphQLDate: bad day 1`] = `"Query error: String is not a valid date string"`;

exports[`dataTypes GraphQLDate: bad month 1`] = `"Query error: String is not a valid date string"`;

exports[`dataTypes GraphQLDateTime 1`] = `"Query error: String is not a valid date time string"`;

exports[`dataTypes GraphQLDateTime: bad hour 1`] = `"Query error: String is not a valid date time string"`;

exports[`dataTypes GraphQLDateTime: bad minute 1`] = `"Query error: String is not a valid date time string"`;

exports[`dataTypes GraphQLDateTime: bad month 1`] = `"Query error: String is not a valid date time string"`;

exports[`dataTypes GraphQLTime 1`] = `"Query error: String is not a valid time string"`;

exports[`dataTypes GraphQLTime: bad hour 1`] = `"Query error: String is not a valid time string"`;

exports[`dataTypes GraphQLTime: bad minute 1`] = `"Query error: String is not a valid time string"`;

exports[`dataTypes GraphQLTime: bad tz hour 1`] = `"Query error: String is not a valid time string"`;
